File: Packages/PaperTimeKit/Sources/LibraryStore/CloudProvider.swift
import Foundation

/// Which sync service a library folder lives in.
///
/// Paper Time does not integrate with any provider's SDK: the library is a
/// plain folder, and whichever service the user already syncs it with does the
/// work. Detection exists only so the interface can say something accurate
/// ("waiting for Google Drive") instead of a generic spinner.
public enum CloudProvider: String, Hashable, Sendable, Codable {
    case iCloudDrive
    case googleDrive
    case dropbox
    case oneDrive
    case box
    case local
    case unknown

    public var displayName: String {
        switch self {
        case .iCloudDrive: "iCloud Drive"
        case .googleDrive: "Google Drive"
        case .dropbox: "Dropbox"
        case .oneDrive: "OneDrive"
        case .box: "Box"
        case .local: "This Device"
        case .unknown: "Cloud Folder"
        }
    }

    public var symbolName: String {
        switch self {
        case .iCloudDrive: "icloud"
        case .local: "internaldrive"
        default: "cloud"
        }
    }

    /// True when files can be evicted and need materialising before reading.
    public var mayEvictFiles: Bool {
        switch self {
        case .local: false
        default: true
        }
    }

    public static func detect(at url: URL) -> CloudProvider {
        if (try? url.resourceValues(forKeys: [.isUbiquitousItemKey]).isUbiquitousItem) == true {
            return .iCloudDrive
        }
        let path = url.path(percentEncoded: false)
        if path.contains("/Mobile Documents/com~apple~CloudDocs") { return .iCloudDrive }
        if path.contains("/CloudStorage/GoogleDrive-") { return .googleDrive }
        if path.contains("/CloudStorage/Dropbox") { return .dropbox }
        if path.contains("/CloudStorage/OneDrive") { return .oneDrive }
        if path.contains("/CloudStorage/Box") { return .box }
        // The iOS document picker hands back File Provider paths that name the
        // provider's bundle identifier.
        let lowered = path.lowercased()
        if lowered.contains("com.google.drive") || lowered.contains("googledrive") {
            return .googleDrive
        }
        if lowered.contains("com.getdropbox") { return .dropbox }
        if lowered.contains("com.microsoft.skydrive") { return .oneDrive }
        if path.hasPrefix(NSHomeDirectory()) || path.hasPrefix("/Users/") { return .local }
        return .unknown
    }
}

File: Portable/src/shared/cloudProvider.ts
/**
 * Which service a library folder is synced by, guessed from where it is.
 *
 * Paper Time speaks to no provider's SDK — the library is a plain folder and
 * whichever service already syncs it does the work. This exists so a row can
 * say where a folder lives: a cloud folder that is slow to answer is waiting,
 * and a folder on this machine that is slow to answer is broken, and the icon
 * is the difference between those two.
 *
 * The Mac's copy is `CloudProvider.swift`. Every desktop puts these folders
 * somewhere else; the answers are the same.
 */
export type CloudProvider =
  | 'iCloudDrive'
  | 'googleDrive'
  | 'dropbox'
  | 'oneDrive'
  | 'box'
  | 'local'

export function providerOf(root: string): CloudProvider {
  // One shape for three desktops: forward slashes, lower case, so a rule is
  // written once instead of three times.
  const path = root.replace(/\\/g, '/').toLowerCase()
  if (path.includes('/mobile documents/com~apple~clouddocs')
    || /(^|\/)iclouddrive(\/|$)/.test(path)) {
    return 'iCloudDrive'
  }
  // Three shapes for one service: the Mac's File Provider folder, the old
  // "Google Drive" folder both desktops used, and the drive Windows mounts —
  // which has no "Google" anywhere in it, only "My Drive" at its root.
  if (path.includes('/cloudstorage/googledrive-')
    || /(^|\/)google[ -]?drive(\/|$)/.test(path)
    || /^[a-z]:\/(my drive|shared drives)(\/|$)/.test(path)) {
    return 'googleDrive'
  }
  if (/(^|\/)dropbox(\/|$)/.test(path)) return 'dropbox'
  // "OneDrive - Acme" is what a work account is called.
  if (/(^|\/)onedrive( -[^/]*)?(\/|$)/.test(path)) return 'oneDrive'
  if (/(^|\/)box(\/|$)/.test(path)) return 'box'
  return 'local'
}

/** The icon a folder's row wears: this machine, or a cloud. */
export function providerIcon(provider: CloudProvider): string {
  if (provider === 'iCloudDrive') return 'icloud'
  if (provider === 'local') return 'internaldrive'
  return 'cloud'
}

export function providerName(provider: CloudProvider): string {
  switch (provider) {
    case 'iCloudDrive': return 'iCloud Drive'
    case 'googleDrive': return 'Google Drive'
    case 'dropbox': return 'Dropbox'
    case 'oneDrive': return 'OneDrive'
    case 'box': return 'Box'
    case 'local': return 'This Device'
  }
}
